### ISSUE TITLE
conditional let being mistaken for never being reassigned

### ISSUE THREAD

**Issue by gabrielcsapo** [eslint/eslint#9922]
**Tell us about your environment**

* ESLint Version:v4.8.0
* Node Version:v8.9.4
* npm Version:5.6.0

parser: prettier

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
let query = {};
if (foo) query.foo = foo;
```

```bash
eslint .
```

**What did you expect to happen?**

Not through an error stating that it should be declared const.

**What actually happened? Please include the actual, raw output from ESLint.**

```
137:11  error  'query' is never reassigned. Use 'const' instead  prefer-const
```

**Comment by gabrielcsapo** [comment]
Started a flame war going to close this out 🕵️‍♂️

**Comment by platinumazure** [comment]
Just for reference: [Here are some examples](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const#Examples) of what `const` does and does not allow you to. Hope this helps!

**Comment by gabrielcsapo** [comment]
@platinumazure the spec is not the issue, the idea that a reference to an object is immutable but the internal references can be, is backwards.

I can say:

```
const data = { world: 'earth', important_number_static: 5000 };
```

Then say:

```
data.world = 'Saturn';
data.important_number_static = 'foo' // :troll:
```

The only way to not exhibit this behavior is by obfuscating the objects values in a class and having setters that set nothing. It was early in the morning when I was like huh, this is weird, when in reality I am the weird one. 😆

**Comment by ljharb** [comment]
Indeed, `const` means "constant reference", not "constant value" :-)
